### PR TITLE
JsDiagGetStackTrace should only give frames which were valid when halt was triggered.

### PR DESCRIPTION
**Pull request by agarwal-sandeep**
JsrtDebugStackFrames::StackFrames should only get frame pointers if the ScriptContext/ProbeContainer had valid stack when halt was triggered. Since user can execute JavaScript code after halt any frames after halt are transient so they should be skipped

### CODE REVIEW

**Comment by agarwal-sandeep** [comment]
@akroshg please take a look

File: lib/Runtime/Debug/ProbeContainer.h [hunk 1]
```diff
@@ -91,6 +91,7 @@ namespace Js
         void Initialize(ScriptContext* pScriptContext);
         void Close();
 
+        bool HaveFramePointers() const { return this->framePointers != nullptr; }
```
Comment (akroshg) [inline lib/Runtime/Debug/ProbeContainer.h]: nit: HasFramePointers

File: lib/Jsrt/JsrtDebuggerObject.cpp [hunk 2]
```diff
@@ -801,24 +801,30 @@ Js::JavascriptArray * JsrtDebugStackFrames::StackFrames(Js::ScriptContext * scri
     tempScriptContext != nullptr && tempScriptContext->IsScriptContextInDebugMode();
         tempScriptContext = tempScriptContext->next)
     {
-        Js::WeakDiagStack * framePointers = tempScriptContext->GetDebugContext()->GetProbeContainer()->GetFramePointers();
-        if (framePointers != nullptr)
+        Js::ProbeContainer* probeContainer = tempScriptContext->GetDebugContext()->GetProbeContainer();
+
+        // Only give frames which were preset when break was triggered as we might have other user (non-library) scripts executed after break
+        if (probeContainer->HaveFramePointers())
```
Comment (akroshg) [inline lib/Jsrt/JsrtDebuggerObject.cpp]: I am not sure if I follow this. We create the stackframe for other scriptcontext only when they were asked for. So when do the other scriptcontext will have their frame pointers populated?

File: lib/Jsrt/JsrtDebugUtils.cpp [hunk 3]
```diff
@@ -27,9 +27,7 @@ void JsrtDebugUtils::AddFileNameOrScriptTypeToObject(Js::DynamicObject* object,
 
         Js::FunctionBody* anyFunctionBody = utf8SourceInfo->GetAnyParsedFunction();
 
-        Assert(anyFunctionBody != nullptr);
-
-        LPCWSTR sourceName = anyFunctionBody->GetSourceName();
+        LPCWSTR sourceName = (anyFunctionBody != nullptr) ? anyFunctionBody->GetSourceName() : Js::Constants::UnknownScriptCode;
```
Comment (akroshg) [inline lib/Jsrt/JsrtDebugUtils.cpp]: Is it an unrelated fix? Is there a possibility that we have utf8sourcecontext without a single function? it should atleast have global function.

File: lib/Jsrt/JsrtDebugManager.cpp [hunk 4]
```diff
@@ -364,8 +364,21 @@ void JsrtDebugManager::CallDebugEventCallback(JsDiagDebugEvent debugEvent, Js::D
...
-        this->debugEventCallback(debugEvent, eventDataObject, this->callbackState);
+        if (isBreak)
+        {
+            // If we are reporting break we should clear all objects after call returns
+            AutoClear autoClear(this);
+
+            // Save the frame address, when asking for stack we will only give stack which is under this address
+            // because host can execute javascript after break which should not be part of stack.
+            void *frameAddress = _AddressOfReturnAddress();
+            scriptContext->GetThreadContext()->GetDebugManager()->SetDispatchHaltFrameAddress(frameAddress);
```
Comment (akroshg) [inline lib/Jsrt/JsrtDebugManager.cpp]: It would be good idea to do SetDispatchHaltFrameAddress(nullptr) after the debugEventCallback call below, so that threadContext->GetDebugManager()->GetDispatchHaltFrameAddress() can return null once the break is over.

**Comment by akroshg** [comment]
Looks good to me
